### PR TITLE
Terraform Import

### PR DESCRIPTION
**Pull request by adamgordonbell**
See also this follow up:
https://github.com/earthly/website/pull/220

It didn't fit in the main article.

### CODE REVIEW

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
+
+- Remove everything via UI and recreate in Terraform.
+  
+  I reached out to Corey, my local Terraform expert and this is the path he recommended:
+
+  > Use the UI once to figure out how to set things up but then tear it down and recreate it in Terraform.  
+
+- Use a 'reverse terraformer'.
+
+  Tools exist that proport to take an existing cloud accounts and import them into Terraform state and generated the Terraform config for them. I did briefly try this approach.
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:310]:
```suggestion
  Tools exist that purport to take existing cloud accounts and import them into Terraform state and generate the Terraform config for them. I did briefly try this approach.
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
+~~~
+
+And I'm all set: My repository is now managed by Terraform.
+
+## Learning From Terraform State with `terraform show`
+
+The Terraform docs are often terse and values in the resource config do not always map one-to-one onto values you'd find the AWS UI. For me, this meant that I couldn't always figure out exactly what I needed to put in the resource configuration to match what I had set up in the UI. Thankfully you can use `terraform show` to retrieve this information.
+
+Let's pretend that I needed my repo to use KMS encryption and to scan image on push. I've already setup these options in AWS but I want to see how to do it AWS.
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:374]:
```suggestion
Let's pretend that I needed my repo to use KMS encryption and to scan an image on push. I've already setup these options in AWS but I want to see how to do it in Terraform.
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 3]
```diff
@@ -0,0 +1,848 @@
...
+  encryption_configuration {
+      encryption_type = "KMS"
+  }
+  image_scanning_configuration {
+        scan_on_push = true
+  }
+}
+~~~
+
+This trick was instrumental in me getting my lambda integration ported over to terraform, but I'll refrain from focusing on it depth. Just remember, if you can't figure out how to configure something in Terraform for some reason, one approach is to set it up in the UI first, and use `terraform import` and `terraform show` to extract a working config.
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:430]:
```suggestion
This trick was instrumental in me getting my lambda integration ported over to terraform, but I'll refrain from focusing on it in depth. Just remember, if you can't figure out how to configure something in Terraform for some reason, one approach is to set it up in the UI first, and use `terraform import` and `terraform show` to extract a working config.
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+## AWS ECR Policy Resource Import
+
+Next up is the ECR Policy. First I created it as a blank resource:
+
+~~~{.groovy caption="main.tf"}
+resource "aws_ecr_repository_policy" "lambda-api" {
+  }
+~~~
+
+Imported is:
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:441]:
```suggestion
Imported with:
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+
+~~~{.bash caption=">_"}
+$ terraform import aws_ecr_repository_policy.lambda-api lambda-api
+~~~
+
+And then with the show trick I extracted the policy I was already using:
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:447]:
```suggestion
And then, with the `terraform show` trick I used above, I extracted the policy I was already using:
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+~~~
+
+And import it by ARN, which I got from the AWS UI:
+
+~~~{.bash caption=">_"}
+$ terraform import aws_acm_certificate.cert arn:...
+$ terraform import aws_api_gateway_domain_name.earthly-tools-com earthly-tools.com
+~~~
+
+Then I need to set up the API, the API State and map it to the domain.
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:587]: Should this be API stage?

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+  source_arn    = "arn:aws:execute-api:us-east-1:459018586415:
+                  ${aws_apigatewayv2_api.earthly-tools-com.id}/*/*/{path+}"
+}
+~~~
+
+Note that `terraform show` returned the current api arn for `source_arn` and I needed to replace that with a path to make this future proof. This is something to keep vigilant for when using `terraform show` or you might end up with resources hardcoded to values that quickly go out of date.
+
+## Terraform Import API Route and Integration
+
+Integrating the lambda with the API was by far the hardest part for me. In AWS, there is a single button click process to hook a lambda up to an API endpoint. Behind the scenes there are a number of separate things happening. But once I figured out the name of all these little resources, it was easy to import and 'terraform show` them to see how they should be configured.
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:679]:
```suggestion
Integrating the lambda with the API was by far the hardest part for me. In AWS, there is a single button click process to hook a lambda up to an API endpoint. Behind the scenes there are a number of separate things happening. But once I figured out the name of all these little resources, it was easy to import and `terraform show` them to see how they should be configured.
```

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 8]
```diff
@@ -0,0 +1,848 @@
...
+│ }
+│ 
+│   with aws_lambda_function.lambda-api,
+│   on main.tf line 126, in resource "aws_lambda_function" "lambda-api":
+│  126: resource "aws_lambda_function" "lambda-api" {
+│ 
+╵
+~~~
+
+In turns out that the lambda needs to have a container in place before being created. So I uncommented the lambda and did a docker push
```
Comment (jalletto) [inline blog/_posts/2022-07-12-terraform-lambda.md:774]:
```suggestion
It turns out that the lambda needs to have a container in place before being created. So I uncommented the lambda and did a docker push
```

**Review by jalletto — CHANGES_REQUESTED** [review]
Looks good. Just a few little things.

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 9]
```diff
@@ -0,0 +1,848 @@
...
+ - terraform import
+---
+<!-- vale on -->
+## From Click Ops To GitOps
+
+Previously I built a [REST API](/blog/aws-lambda-api-proxy/), deployed it into a container and got it all running on AWS as a Lambda. But setting this up involved just clicking around in AWS and occasionally using the AWS CLI.
+
+Today, I'm going to port that whole setup to Terraform so that its easier to manage, reproduce, and make changes to. ( These are the purported benefits of Terraform, or so my coworker Corey tells me. I've never used it before so I'll be learning as I go. )
+
+**Read this to learn about Terraform from a noob's perspective. The specifics of my infrastructure import may not be relevant to you, but hopefully my thinking process and take-aways will be.**
```
[the remark this answers is not in the document]
Comment (adamgordonbell) [reply blog/_posts/2022-07-12-terraform-lambda.md:20]: yeah, makes sense. 

There are probably less explicit ways to do it, but I think of it as saying "Here is why you should  (or shouldn't) read this" somewhere right at the beginning.

File: blog/_posts/2022-07-12-terraform-lambda.md (hunk 6 above)
Comment (adamgordonbell) [reply blog/_posts/2022-07-12-terraform-lambda.md:587]: Yes, nice catch.

File: blog/_posts/2022-07-12-terraform-lambda.md [hunk 10]
```diff
@@ -0,0 +1,848 @@
...
+
+To debug this you need to understand that the aws provider reads from `~/.aws/confg` and `~/.aws/credentials`, the same way the `aws` cli does. If you have credentials for more than one AWS account available, you can use `AWS_PROFILE` to configure the AWS provider. In my case I do this like so:
+
+~~~{.bash caption=">_"}
+$ export AWS_PROFILE=earthly-dev
+~~~
+
+It's also possible to configure the profile in terraform by adding it under provider configuration:
+
+~~~{.groovy caption="main.tf"}
```
[the remark this answers is not in the document]
Comment (adamgordonbell) [reply blog/_posts/2022-07-12-terraform-lambda.md]: Somehow this wasn't happening once I made other changes. Hopefully it doesn't come back :)

**Review by jalletto — APPROVED** [review]

**Pull request merged**